common/persistence/cassandra/queue_store_v2.go

```diff
@@ -0,0 +1,356 @@
+	templateGetLastMessageIDQueryV2 = `SELECT message_id FROM queue WHERE queue_type=? and queue_name=? and queue_partition=? ORDER BY message_id DESC LIMIT 1`
+	templateDeleteMessagesQueryV2   = `DELETE FROM queue WHERE queue_type = ? and queue_name = ? and message_id > ? and message_id <= ?`
+	templateGetMessagesQueryV2      = `SELECT message_id, message_payload, message_encoding FROM queue WHERE queue_type = ? and queue_name = ? and partition = ? and message_id > ? LIMIT ?`
+
+	templateInsertQueueMetadataQueryV2          = `INSERT INTO queue_metadata_v2 (queue_type, queue_name, metadata_payload, metadata_encoding, version) VALUES(?, ?, ?, ?, ?) IF NOT EXISTS`
+	templateGetQueuesListQueryV2                = `SELECT queue_name FROM queue_metadata_v2 WHERE queue_type=?`
+	templateGetCurrentAckLevelAndVersionQueryV2 = `SELECT metadata_payload, metadata_encoding, version FROM queue_metadata_v2 WHERE queue_type = ? and queue_name = ?`
+	templateUpdateQueueMetadataQueryV2          = `UPDATE queue_metadata SET metadata_payload = ?, metadata_encoding= ?, version = ? WHERE queue_type = ? and queue_name = ? IF version = ?`
+
+	EmptyPartition = 0
+)
+
+type (
+	QueueStoreV2 struct {
+		session gocql.Session
+		logger  log.Logger
+	}
+
+	// TODO: add a proto at some point
+	QueueV2MetadataPayload struct {
+		AckLevel int64
+	}
+
+	// both MessageID and Pagetoken are required to reconstrcut the query
+	// This is encoded to []byte and sent over the wire to the client
+	// Initially both are set to nil
+	InternalReadMessagePageToken struct {
+		MessageID int64
+		PageToken []byte
+	}
+)
+
+func NewQueueStoreV2(
+	queueType persistence.QueueV2Type,
+	session gocql.Session,
+	logger log.Logger,
+) (persistence.QueueV2, error) {
+	return &QueueStoreV2{
+		session: session,
+		logger:  logger,
+	}, nil
+}
+
+func (q *QueueStoreV2) CreateQueue(
+	ctx context.Context,
+	request persistence.InternalCreateQueueRequest,
+) (*persistence.InternalCreateQueueResponse, error) {
+	version := 0
+	blob, err := convertQueueV2MetadataToBlob(&QueueV2MetadataPayload{AckLevel: 0})
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert initial queue metadata record: %w, Type: %v", err, request.QueueType)
+	}
+	query := q.session.Query(templateInsertQueueMetadataQueryV2,
+		request.QueueType,
+		request.QueueName,
+		blob.Data,
+		blob.EncodingType.String(),
+		version,
+	).WithContext(ctx)
+	_, err = query.MapScanCAS(make(map[string]interface{}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert initial queue metadata record: %w, Type: %v", err, request.QueueType)
+	}
+	// it's ok if the query is not applied, which means that the record exists already.
+	return &persistence.InternalCreateQueueResponse{}, nil
+}
+
+func (q *QueueStoreV2) EnqueueMessage(ctx context.Context, request persistence.InternalEnqueueMessageRequest) (*persistence.InternalEnqueueMessageResponse, error) {
```

Comment:
For cassandra, we need to make sure this concurrency for this method is 1 on a host and tryEnqueue should do backoff while holding the concurrency token when it encounters conditional failure. Please add TODO for that.